storage/postgres/tree_storage.go

```diff
@@ -15,19 +15,378 @@
 package postgres
 
 import (
+	"context"
 	"database/sql"
+	"encoding/base64"
+	"fmt"
+	"runtime/debug"
+	"strings"
+	"sync"
 
 	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/cache"
+	"github.com/google/trillian/storage/storagepb"
 )
 
+const (
+	placeholderSQL        = "<placeholder>"
+	insertSubtreeMultiSQL = `INSERT INTO subtree(tree_id, subtree_id, nodes, subtree_revision) ` + placeholderSQL
+	selectSubtreeSQL      = `
+		SELECT x.subtree_id, x.max_revision, subtree.nodes
+		FROM (
+			SELECT n.subtree_id, max(n.subtree_revision) AS max_revision
+			FROM subtree n
+			WHERE n.subtree_id IN (` + placeholderSQL + `) AND
+			n.tree_id = ? AND n.subtree_revision <= ?
+			GROUP BY n.subtree_id
+		) AS x
+		INNER JOIN subtree
+		ON subtree.subtree_id = x.subtree_id
+		AND subtree.subtree_revision = x.max_revision
+		AND subtree.tree_id = ?`
+)
+
+// pgTreeStorage contains the pgLogStorage implementation.
+type pgTreeStorage struct {
+	db *sql.DB
+
+	// Must hold the mutex before manipulating the statement map. Sharing a lock because
+	// it only needs to be held while the statements are built, not while they execute and
+	// this will be a short time. These maps are from the number of placeholder '$#'
+	// in the query to the statement that should be used.
+	statementMutex sync.Mutex
+	statements     map[string]map[int]*sql.Stmt
+}
+
 // OpenDB opens a database connection for all PG-based storage implementations.
 func OpenDB(connStr string) (*sql.DB, error) {
 	db, err := sql.Open("postgres", connStr)
 	if err != nil {
-		// Don't log conn str as it could contain credentials
+		// Don't log conn str as it could contain credentials.
 		glog.Warningf("Could not open Postgres database, check config: %s", err)
 		return nil, err
 	}
 
 	return db, nil
 }
+
+func newTreeStorage(db *sql.DB) *pgTreeStorage {
+	return &pgTreeStorage{
+		db:         db,
+		statements: make(map[string]map[int]*sql.Stmt),
+	}
+}
+
+// statementSkeleton contains the structure of a query to create.
+type statementSkeleton struct {
```

Comment:
I assume this gets used a lot more in future PRs, thereby justifying its existence? It looks rather complicated for the two uses that appear in this PR. All of this string substitution into SQL and use of variables as printf format strings makes me a little uneasy!

Reply:
Well if @AlCutter is happy with it, I don't want to block you any longer. I'll try and find the time to have a think about whether there's a better way of going about this and come back to it.

Reply:
Thanks Rob, I just created #1369 to track this.